// unleash-server/dist/lib/middleware/no-authentication.js
import NoAuthUser from '../types/no-auth-user.js';
import { ApiTokenType } from '../types/model.js';
import { DEFAULT_ENV } from '../util/index.js';
import ApiUser from '../types/api-user.js';
import * as permissions from '../types/permissions.js';
// eslint-disable-next-line
function noneAuthentication(baseUriPath, app) {
    app.use(`${baseUriPath || ''}/api/admin/`, (req, res, next) => {
        if (!req.user) {
            req.user = new NoAuthUser();
        }
        next();
    });
}
export function noApiToken(baseUriPath, app) {
    app.use(`${baseUriPath}/api/frontend`, (req, res, next) => {
        if (!req.headers.authorization && !req.user) {
            req.user = new ApiUser({
                tokenName: 'unknown',
                permissions: [permissions.FRONTEND],
                projects: ['*'],
                environment: DEFAULT_ENV,
                type: ApiTokenType.FRONTEND,
                secret: 'unknown',
            });
        }
        next();
    });
    app.use(`${baseUriPath}/api/client`, (req, res, next) => {
        if (!req.headers.authorization && !req.user) {
            req.user = new ApiUser({
                tokenName: 'unknown',
                permissions: [permissions.CLIENT],
                projects: ['*'],
                environment: DEFAULT_ENV,
                type: ApiTokenType.CLIENT,
                secret: 'unknown',
            });
        }
        next();
    });
}
export default noneAuthentication;
