// unleash-server/dist/lib/logger.js
import pkg from 'log4js';
const { configure, getLogger } = pkg;
export var LogLevel;
(function (LogLevel) {
    LogLevel["debug"] = "debug";
    LogLevel["info"] = "info";
    LogLevel["warn"] = "warn";
    LogLevel["error"] = "error";
    LogLevel["fatal"] = "fatal";
})(LogLevel || (LogLevel = {}));
export function getDefaultLogProvider(logLevel = LogLevel.error) {
    configure({
        appenders: {
            console: { type: 'console' },
        },
        categories: {
            default: { appenders: ['console'], level: logLevel },
        },
    });
    return getLogger;
}
function validate(isValid, msg) {
    if (!isValid) {
        throw new TypeError(msg);
    }
}
export function validateLogProvider(provider) {
    validate(typeof provider === 'function', 'Provider needs to be a function');
    const logger = provider('unleash:logger');
    validate(typeof logger.debug === 'function', 'Logger must implement debug');
    validate(typeof logger.info === 'function', 'Logger must implement info');
    validate(typeof logger.warn === 'function', 'Logger must implement warn');
    validate(typeof logger.error === 'function', 'Logger must implement error');
}
