// unleash-server/dist/lib/features/playground/feature-evaluator/helpers.js
export function createFallbackFunction(name, context, fallback) {
    const createEvalResult = (enabled) => ({
        result: enabled,
        strategies: [],
    });
    if (typeof fallback === 'function') {
        return () => createEvalResult(fallback(name, context));
    }
    if (typeof fallback === 'boolean') {
        return () => createEvalResult(fallback);
    }
    return () => createEvalResult(false);
}
export function resolveContextValue(context, field) {
    if (context[field]) {
        return context[field];
    }
    if (context.properties?.[field]) {
        return context.properties[field];
    }
    return undefined;
}
export function safeName(str = '') {
    return str.replace(/\//g, '_');
}
export function getDefaultStrategy(featureName) {
    return {
        name: 'flexibleRollout',
        constraints: [],
        disabled: false,
        parameters: {
            rollout: '100',
            stickiness: 'default',
            groupId: featureName,
        },
    };
}
function resolveGroupId(defaultStrategy, featureName) {
    const groupId = defaultStrategy?.parameters?.groupId !== ''
        ? defaultStrategy.parameters?.groupId
        : featureName;
    return groupId || '';
}
export function getProjectDefaultStrategy(defaultStrategy, featureName) {
    return {
        ...defaultStrategy,
        parameters: {
            ...defaultStrategy.parameters,
            groupId: resolveGroupId(defaultStrategy, featureName),
        },
    };
}
